### DISCUSSION TITLE
Dependencies with yield not working with StreamingResponse

### DISCUSSION THREAD

**Discussion by hasansustcse13** [fastapi/fastapi#11444] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class MyClass:
    def release(self):
        print('close')


async def get_instance():
    obj = MyClass()
    try:
        print('before')
        yield obj
        print('after')
    finally:
        print('Call after streaming completed')
        obj.release()


async def stream_test():
    for i in range(50):
        yield f"event: message\ndata: {i}\n\n"
        await asyncio.sleep(.1)
    print('streaming completed')


@router.get('/chat', name='Chat with RAG')
async def rag_chat_async(instance: Annotated[MyClass, Depends(get_instance)]):
    return StreamingResponse(stream_test(), media_type='text/event-stream')
```

### Description

After updating to `0.110.0`, Dependencies with yield not working. Finally block has been executed before the streaming completed.
**Expected Behavior**: Finally block should execute after the streaming completed. 

### Operating System

Windows

### FastAPI Version

0.110.0

### Pydantic Version

2.6.3

### Python Version

3.11

### Additional Context

starlette==0.36.3

**Comment by hasansustcse13** [comment]
@tiangolo could you please share your thoughts here?

**Comment by pythrick** [comment]
I've just found out this issue as well. This behavior was introduced in version 0.106, when they changed how dependencies with yield work: https://fastapi.tiangolo.com/release-notes/#01060.

Here's my snipped code with similar issue:

```python

class TempClient:
    async def __aenter__(self):
        self.x = "a"
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        self.x = "b"

    async def get_x(self):
        yield self.x
        await asyncio.sleep(1)
        yield self.x
        await asyncio.sleep(1)
        yield self.x
        await asyncio.sleep(1)


async def get_tmp_client():
    async with TempClient() as tmp_client:
        yield tmp_client


@app.get("/logs")
async def get_stream_logs(tmp_client: TempClient = Depends(get_tmp_client)):
    return StreamingResponse(
        tmp_client.get_x(),
        media_type="text/plain"
    )
```

Output with version 0.105.0:
```shell
curl -X 'GET' \                                                                                                           
  'http://localhost:5000/logs' \
  -H 'accept: application/json' --no-buffer
aaa%
```

Output with version 0.106.0:
```shell
 curl -X 'GET' \                                                                                                           
  'http://localhost:5000/logs' \
  -H 'accept: application/json' --no-buffer
bbb%
````

I believe they have though that this would only impact Background tasks, without measuring the impact on Streaming responses.
Instantiating a dependency for each streaming chunk doesn't seems the right approach.

**Comment by miguelvalente** [comment]
@pythrick Thanks for pointing out the version :)

**Comment by sanderbollen-clockworks** [comment]
Are there any updates on this? What is the recommended approach for StreamingResponses going forward?

**Reply by rustonaut** [reply]
Until any fixes/changes are done upstream options are quite limited:

1. you can moving the `with`-block/context manager of the dependency into the generator function you pass to `StreamingResponse` is pretty much the only convenient+easy approach.  But that falls apart really fast.

2. you can not use with blocks and more manual management, but that is likely to not work well in a lot of ways

3. have some way to "move out" the resource guarded by a context manager (moving it into another context manager in the generator) but that is in many ways not how context managers are supposed to work and might lead to a lot of mayhem. For example if dependencies depend on dependencies or if  a open context managers is accidentally moved into adjacent tasks with different task local data and similar . I.e. it's a complete nightmare through might happen to work in a some contexts :see_no_evil:  

So I think this is really something which should be fixed/changed upstream in some way as there is no good workaround downstream  as far as I can tell.

Through I might also have missed something.

**Reply by mikkelam** [reply]
In my team we've decided to stick with version `0.105.0` until it's resolved

**Comment by AHTOOOXA** [comment]
StreamingResponse basically calls generator func after all the dependencies are closed. It's such a weird and counter intuitive behaviour, causes me database sessions leaks which was a 8 hour pain in the ass to find with no observable solution

**Comment by jaysephjw** [comment]
@tiangolo any movement on this? 

Our team lost a lot of time on this investigating why our service was leaking db connections.

**Comment by YuriiMotov** [comment]
Related answer from tiangolo: https://github.com/fastapi/fastapi/issues/11143#issuecomment-2414188820

So, it will probably be fixed one day.
For now the solution is to avoid using the results of dependencies with `yield` in streaming response generators. The same approach as with `BackgroundTasks`: 

> now you should create the resources for background tasks inside the background task itself, and use internally only data that doesn't depend on the resources of dependencies with yield.
> 
> For example, instead of using the same database session, you would create a new database session inside of the background task, and you would obtain the objects from the database using this new session. And then instead of passing the object from the database as a parameter to the background task function, you would pass the ID of that object and then obtain the object again inside the background task function.

Source: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#background-tasks-and-dependencies-with-yield-technical-details

**Reply by baggiponte** [reply]
uh interesting, but does that make sense for a streaming response too? I think creating a resource with every yield adds up latency.

**Reply by YuriiMotov** [reply]
You don't need to create resource on every yield, you should create resource when you start streaming:
```py
async def my_stream_response(sessionmaker):
    with sessionmaker() as session:
        for _ in range(10):
            yield "something"
```

**Reply by baggiponte** [reply]
that's right, thanks! :)

**Comment by AHTOOOXA** [comment, marked as answer]
hey guys i think the problem with streaming responses has been resolved in https://github.com/fastapi/fastapi/releases/tag/0.118.0

**Reply by baggiponte** [reply]
can confirm, we run 0.118 in prod and the fix works.
